### PR TITLE
Logging improvement

### PR DESCRIPTION
**Pull request by squirrelsc**
See details in commit

1. Support log path for environment and node. They can save useful information in their folders. The content can be used for troubleshooting cross test cases.
2. Save newly downloaded serial log every time.
3. Make get information more stable.
4. Parse wala version from serial log too.

### CODE REVIEW

File: lisa/environment.py [hunk 1]
```diff
@@ -146,6 +148,9 @@ def __init__(self, is_predefined: bool, warn_as_error: bool) -> None:
         self.warn_as_error = warn_as_error
         self._default_node: Optional[Node] = None
         self._log = get_logger("env", self.name)
+        # delay construct it until it's used, because the environment name is not set
```
Comment (LiliDeng) [inline lisa/environment.py]: I don't understand the context for this comments.

File: lisa/features/serial_console.py [hunk 2]
```diff
@@ -58,6 +63,26 @@ def invalidate_cache(self) -> None:
         )
         self._cached_console_log = None
 
+    def get_matched_str(self, pattern: Pattern[str]) -> str:
```
Comment (glima) [inline lisa/features/serial_console.py:67]: Same name as the imported one, so confusing... Also, amazed that our make check does not complain about this empty line here.

File: lisa/environment.py [hunk 3]
```diff
@@ -195,6 +200,23 @@ def create(
...
+            self._log_path = (
+                constants.RUN_LOCAL_PATH
+                / "environments"
+                / f"{get_datetime_path()}-{self.name}"
+            )
+            if self._log_path.exists():
+                raise LisaException(
+                    "Conflict environment path, "
+                    "make sure every environment have an unique log path"
+                    f"'{self._log_path}'"
```
Comment (glima) [inline lisa/environment.py:219]: Maybe "Conflicting log path detected: {path_var_here}. Make sure LISA invocations have individual runtime paths and don't issue concurrent invocations."?

File: lisa/node.py [hunk 4]
```diff
@@ -191,6 +198,29 @@ def support_sudo(self) -> bool:
     def is_connected(self) -> bool:
         return self._shell is not None and self._shell.is_connected
 
+    @property
+    def local_log_path(self) -> Path:
+        if not self._local_log_path:
```
Comment (glima) [inline lisa/node.py:203]: Why not return early on the good case and avoid the indentation?

File: lisa/node.py [hunk 5]
```diff
@@ -191,6 +198,29 @@ def support_sudo(self) -> bool:
...
+                if self.index:
+                    index = self.index
+                else:
+                    index = randint(0, 10000)
+                path_name = f"node-{index}"
+            self._local_log_path = base_path / path_name
+            if self._local_log_path.exists():
+                raise LisaException(
+                    f"Conflict node path, make sure every node have an unique log path"
+                    f"'{self._local_log_path}'"
```
Comment (glima) [inline lisa/node.py:219]: Ditto for message.

File: lisa/sut_orchestrator/azure/platform_.py [hunk 6]
```diff
@@ -117,9 +117,15 @@
 )
 KERNEL_VERSION_PATTERN = re.compile(r"Linux version (?P<kernel_version>.+?) ", re.M)
 
+# 2021/03/31 00:05:17.431693 INFO Daemon Azure Linux Agent Version:2.2.38
+WALA_VERSION_PATTERN = re.compile(
+    r"Azure Linux Agent Version:(?P<wala_version>.+?)[\n\r]", re.M
```
Comment (glima) [inline lisa/sut_orchestrator/azure/platform_.py:125]: is wala some outdated name? can we stick to the current one, if so?

Comment (squirrelsc) [reply lisa/sut_orchestrator/azure/platform_.py:125]: Not sure what's current name, but it's used in database and not exposed to runbook. Let's keep as is.

File: lisa/environment.py (hunk 1 above)
Comment (squirrelsc) [reply lisa/environment.py]: Improved the message, hope it's clearer

File: lisa/features/serial_console.py (hunk 2 above)
Comment (squirrelsc) [reply lisa/features/serial_console.py:67]: They perform the same functionality. The called one is for any string, this is in the SerialConsole class, only for serial log.
Remove the empty line to keep consistence.

File: lisa/environment.py (hunk 3 above)
Comment (squirrelsc) [reply lisa/environment.py:219]: Revised some. The concurrent invocations are allowed, and the env name is unique. Hope this situation won't happen.

File: lisa/node.py (hunk 4 above)
Comment (squirrelsc) [reply lisa/node.py:203]: In some coding guidline, it recommends to single exit entry in each function. I'm struggling sometime, but following it.

File: lisa/node.py (hunk 5 above)
Comment (squirrelsc) [reply lisa/node.py:219]: revised.

File: lisa/sut_orchestrator/azure/platform_.py (hunk 6 above)
Comment (glima) [reply lisa/sut_orchestrator/azure/platform_.py:125]: @simonxiaoss any clue?

**Pull request merged**